### PR TITLE
msgr/async: V1 protocol refactor 

### PR DESCRIPTION
**Pull request by rjfd**
The main idea of the refactoring in this PR is to move all V1 protocol logic and state from the `AsyncConnection.cc` to its own class, and in the `AsyncConnection` we store a pointer to the protocol implementation. That way we can implement the V2 protocol and easily switch between V1 and V2 in the `AsyncConnection` class by instantiation the respective implementation.

### CODE REVIEW

**Comment by rjfd** [comment]
@liewegas @yuyuyu101 I removed the WIP from the title and the PR is now ready for reviewing and to run integration tests.

File: src/msg/async/Protocol.h [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+
+  virtual ~ProtocolV1();
+
+  virtual void connect();
+  virtual void accept();
+  virtual bool is_connected();
+  virtual void stop();
+  virtual void fault();
+  virtual void send_message(Message *m);
+  virtual void send_keepalive();
```
[the remark this answers is not in the document]
Comment (rjfd) [reply src/msg/async/Protocol.h]: done

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2018-09-11_18:22:38-rados-wip-sage3-testing-2018-09-10-1637-distro-basic-smithi/

lots of failures to look through...

**Comment by gregsfortytwo** [comment]
@rjfd, do you think that latest patch should have resolved all the issues that testing turned up, or is there more coming?

**Comment by rjfd** [comment]
@gregsfortytwo @liewegas rados:thrash still green:
http://pulpito.ceph.com/rdias-2018-09-18_11:02:23-rados:thrash-wip-rdias-testing-distro-basic-smithi/

**Comment by liewegas** [comment]
Looks good in qa too.  @yuyuyu101 any concerns before we merge?

@rjfd we grew a small conflict, can you rebase?

**Comment by rjfd** [comment]
@liewegas done rebase

**Comment by tchaikov** [comment]
@cyx1231st in case you are interest.

**Comment by liewegas** [comment]
retest this please

File: src/msg/async/Protocol.cc [hunk 2]
```diff
@@ -0,0 +1,2456 @@
...
+  if (connect_msg.protocol_version != reply.protocol_version) {
+    send_connect_message_reply(CEPH_MSGR_TAG_BADPROTOVER, reply,
+                               authorizer_reply);
+    return;
+  }
+
+  // require signatures for cephx?
+  if (connect_msg.authorizer_protocol == CEPH_AUTH_CEPHX) {
+    if (connection->peer_type == CEPH_ENTITY_TYPE_OSD ||
+        connection->peer_type == CEPH_ENTITY_TYPE_MDS) {
```
Comment (runsisi) [inline src/msg/async/Protocol.cc]: hi @rjfd why `MON` and `MGR` are not in this list ? thanks.

**Pull request merged**